### PR TITLE
8235491: Tree/TableView: implementation of isSelected(int) violates contract

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
1. reword SelectionModel.isSelected(int) javadoc, removing incorrect statement "Is functionally equivalent to calling <code>getSelectedIndices().contains(index)</code>." 
2. reimplement TableView.TableViewSelectionModel.isSelected(int) method to return true when at least one cell in *any* column is selected on the given row (was: *all* columns)
3. modified TreeTableRow.updateSelection() to use the right isSelected() method
4. updated tests for Tree/TableView

NOTE: proposed change alters semantics of isSelected(int) method (in the right direction, in my opinion).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8235491](https://bugs.openjdk.org/browse/JDK-8235491): Tree/TableView: implementation of isSelected(int) violates contract
 * [JDK-8290741](https://bugs.openjdk.org/browse/JDK-8290741): Tree/TableView: implementation of isSelected(int) violates contract (**CSR**)

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Jeanette Winzenburg](https://openjdk.org/census#fastegal) (@kleopatra - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/839/head:pull/839` \
`$ git checkout pull/839`

Update a local copy of the PR: \
`$ git checkout pull/839` \
`$ git pull https://git.openjdk.org/jfx pull/839/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 839`

View PR using the GUI difftool: \
`$ git pr show -t 839`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/839.diff">https://git.openjdk.org/jfx/pull/839.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Because there is a spec change, and a behavioral change, this needs a CSR. What are the implications of this change on applications?

/reviewers 2
/csr

**Comment by andy-goryachev-oracle** [comment]
I think the behavior of utility method SelectionModel.isSelected(int) was poorly defined in the context of TableView, and/or incorrect in case of cell selection enabled.

People who used TableView.isSelected(int, TableColumn|TableColumnBase) will be ok, but those who assumed (arguable incorrect) implementation may need to revisit their code, and only in the case when cell selection is enabled.

**Comment by kevinrushforth** [comment]
OK. When you create the CSR, you can add a note about possible compatibility impact. Seems it should be low.

**Comment by kevinrushforth** [comment]
@kleopatra Would you have time / interest to review this?

**Comment by andy-goryachev-oracle** [comment]
Kevin:
Proposed CSR - JDK-8290741
what would be the "correct fix version"?

**Comment by kevinrushforth** [comment]
> Kevin: Proposed CSR - JDK-8290741 what would be the "correct fix version"?

`openjfx20`

**Review by aghaisas — CHANGES_REQUESTED** [review]
1. I see that you have made below change for cell selection mode -

- Before : `isSelected(int)` method used to return `True` only if **all** of the cells in that row were selected.
- After : `isSelected(int)` method returns `True` if **any** of the cells in that row are selected.

     Now, if someone wants to know - while in Cell selection mode, whether all of the cells in a particular row are selected? then they must use another already existing API `isSelected(int row, TableColumn<S,?> column)` with `column` parameter as `null`.
This is important and should be captured somewhere as recommendation.

2. Similar change needs to be made for `TreeTableView` as well
3. I have listed a minor comment on coding style

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java [hunk 1]
```diff
@@ -2813,8 +2813,20 @@ private void quietClearSelection() {
             stopAtomic();
         }
 
-        @Override public boolean isSelected(int index) {
-            return isSelected(index, null);
+        @Override
+        public boolean isSelected(int index) {
+            final boolean isCellSelectionEnabled = isCellSelectionEnabled();
+            if (isCellSelectionEnabled) {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java:2819]: +1

[the remark this answers is not in the document]
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java]: >     1. javadoc for TableSelectionModel.isSelected(int, TablecolumnBase) already describes the logic:
Yes. I am aware of this documentation. I was thinking of making it clear for the users if they see their application breaks due to this change. I see that you have captured this in "Compatibility Impact" section of the CSR.


>     2. TreeTableView.TreeTableViewSelectionModel extends TableSelectionModel, so the changes affect both.
I see that there are overridden methods `public boolean isSelected(int index)` and `public boolean isSelected(int row, TableColumnBase<TreeItem<S>,?> column)` in class `TreeTableViewArrayListSelectionModel` present in TreeTableView.java. Hence, I think that the change that you have made in `TableView.java - isSelected(int index)` method should also be made in `TreeTableView.java - isSelected(int index)` as well.


>     3. good point, fixed.
Thanks!

[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java]: 1. Would you suggest possible clarification please?  The javadoc for SelectionModel.isSelected(int) is changed to say
```
Convenience method to inform if the given index is currently selected
     * in this SelectionModel.  It will return true when at least one cell
     * is selected in the specified row index.
```

2. you are right!  fixed.

**Comment by aghaisas** [comment]
> 1. Would you suggest possible clarification please?

As you have mentioned the impact of this change & using alternate API in "Compatibility Impact" section of the CSR - we are good! No additional change is required.

One minor issue - the description of this PR starts with `...ontract` - looks like some typo.

**Comment by andy-goryachev-oracle** [comment]
> One minor issue - the description of this PR starts with `...ontract` - looks like some typo.

github truncated the title and placed the remainder into PR description, I just left it there.  fixed.

File: modules/javafx.controls/src/main/java/javafx/scene/control/SelectionModel.java [hunk 2]
```diff
@@ -199,8 +199,8 @@ public SelectionModel() { }
 
     /**
      * <p>Convenience method to inform if the given index is currently selected
-     * in this SelectionModel. Is functionally equivalent to calling
-     * <code>getSelectedIndices().contains(index)</code>.
+     * in this SelectionModel.  It will return true when at least one cell
+     * is selected in the specified row index.
      *
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/SelectionModel.java:204]: Well, I would go a step (actually 2 steps :) further:

A:

- the doc error was that a method in the base class specifies its behavior with api from a sub: can't and should be done because it doesn't know its subs plus there are subs that don't have such api (like SingleSelection model). 
- the fix is to replace the description with a similar error: there is no notion of "cells in a row" in the base and also not in many of its subs

I would suggest to simply remove the old sentence without replacing it with anything: it has the notion of multiple selections without actually api to access those multiples, selected could stand for its own and might be fleshed out by subclasses.

B: 

we need a CSR anyway, so might cleanup a bit further: this method is _not_ a convenience method, quite on the contrary - for clients coding against this base class, it's the _only_ means to get hold of the selected indices (by looping through all indices in the valid range and calling it)

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java [hunk 3]
```diff
@@ -2813,8 +2813,19 @@ private void quietClearSelection() {
...
+                for (int col = 0; col < columnCount; col++) {
+                    if (selectedCellsMap.isSelected(index, col)) {
+                        return true;
+                    }
+                }
+                return false;
+            } else {
+                return selectedCellsMap.isSelected(index, -1);
+            }
         }
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java:2829]: wondering why you distinguish between not/ cellSelection? This method is all about row selection (that is the dimension available in base/multiple selection) - shouldn't it behave the same way, independent of whether the second - cell - dimension is turned on? If so, we could simply delegate to super

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java [hunk 4]
```diff
@@ -3165,8 +3165,19 @@ private void quietClearSelection() {
...
+                for (int col = 0; col < columnCount; col++) {
+                    if (selectedCellsMap.isSelected(index, col)) {
+                        return true;
+                    }
+                }
+                return false;
+            } else {
+                return selectedCellsMap.isSelected(index, -1);
+            }
         }
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:3181]: same comment as tableView - just to not forget :)

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewSelectionModelImplTest.java [hunk 5]
```diff
@@ -172,7 +172,7 @@ private TablePosition pos(int row, TableColumn<String,?> col) {
         assertFalse(model.isSelected(1, col0));
         assertFalse(model.isSelected(3, col0));
         assertFalse(cells(model), model.isSelected(3, null));
-        assertFalse(model.isSelected(3));
+        assertTrue(model.isSelected(3));
         assertEquals(1, model.getSelectedCells().size());
     }
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewSelectionModelImplTest.java:178]: do we have tests isSelected(index) for all combinations of single/multiple/cellSelection? If not, now might be a good time to add them :)

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewSelectionModelImplTest.java [hunk 6]
```diff
@@ -134,7 +134,7 @@ private TreeTablePosition pos(int row, TreeTableColumn<String,?> col) {
         assertFalse(model.isSelected(1, col0));
         assertFalse(model.isSelected(3, col0));
         assertFalse(cells(model), model.isSelected(3, null));
-        assertFalse(model.isSelected(3));
+        assertTrue(model.isSelected(3));
         assertEquals(1, model.getSelectedCells().size());
     }
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewSelectionModelImplTest.java:139]: same comment as TableView :)

**Review by kleopatra — CHANGES_REQUESTED** [review]
Basically good fix, left a couple of change suggestions/questions inline :)

**Comment by kleopatra** [comment]
forgot: will not be near my IDE until next Monday - then I'll run the tests and see for myself :)

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java (hunk 3 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java:2829]: Good question!
The implementation is lifted from its isSelected(int,Tree/TableColumn) sibling with the logic altered for the case of enabled cell selection.
Let me check if the logic can be delegated to superclass.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewSelectionModelImplTest.java (hunk 5 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewSelectionModelImplTest.java:178]: a *very* good point!
should definitely cover all the scenarios.

**Comment by andy-goryachev-oracle** [comment]
@kleopatra : 
Various scenarios are pretty well covered by the existing tests.  Just in case, I've added tests to exercise the model with cell selection enabled.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java]: @aghaisas 

> Yes. I am aware of this documentation. I was thinking of making it clear for the users if they see their application breaks due to this change.

If any application breaks, it's its own fault: they coded against the implementation of TableXXSelectionModels which was (incorrectly) `isSelected(int) == isSelected(int, null)`.

That might be due to the specification mess, though: 

- the old spec on `SelectionModel.isSelected(int)` was incorrectly arguing with api on MultipleSelectionModel, that is _Is functionally equivalent to calling getSelectedIndices().contains(index)_
- that invariant actually belongs into `MultipleSelectionModel.isSelected(int)` which has no specialized spec
- the method is first implemented in the (unfortunately hidden) `MultipleSelectionModelBase.isSelected(int)` fulfilling the constraint but also without spec.

To clean this up completely, we could also change the of MultipleSelectionModel and move the old `isSelected(int) == getSelectedIndices().contains(int)` into the spec of its isSelected. Probably could be done in a follow-up issue (or added to one of the open doc errors around selection models).

What do you think?

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java]: > To clean this up completely, we could also change the of MultipleSelectionModel and move the old `isSelected(int) == getSelectedIndices().contains(int)` into the spec of its isSelected. Probably could be done in a follow-up issue (or added to one of the open doc errors around selection models).
> 
> What do you think?

I like this idea. Either filing a new bug or fold into an existing one seems fine.

**Comment by kevinrushforth** [comment]
The doc changes look fine to me, since I agree that further clarification of `MultipleSelectionModel.isSelected` can be done in a follow-up.

**Comment by kleopatra** [comment]
a bit confused about the [csr](https://bugs.openjdk.org/browse/JDK-8290741) - shouldn't that be focused entirely on SelectionModel.isSelected)? That's where we clarify the contract - all changes to TableXXSelectionModels are implementation changes, fixing their contract violation. So I would expect these not to be mentioned at all in the csr.

BTW: isSelected is _not_ a convenience method - that's already changed here, all mentions of its being so should be removed from the csr as well, IMO :)

**Comment by kleopatra** [comment]
> @kleopatra : Various scenarios are pretty well covered by the existing tests. 

good - we seem to be lucky :)

**Comment by kevinrushforth** [comment]
> a bit confused about the csr - shouldn't that be focused entirely on SelectionModel.isSelected)? That's where we clarify the contract - all changes to TableXXSelectionModels are implementation changes, fixing their contract violation. So I would expect these not to be mentioned at all in the csr.

The specification section of the CSR should (and does) list only the spec change to `SelectionModel.isSelected`. Since there could be an impact to applications that rely on the current behavior, it also seems reasonable to list those in the Solution section (as well as the Compatibility concerns section). I think that the Summary should start by listing the spec change to `SelectionModel.isSelected`, but it seems useful to also say that the implementation of `TableXXSelectionModel` is changing as a result.

> BTW: isSelected is not a convenience method - that's already changed here, all mentions of its being so should be removed from the csr as well, IMO :)

Agreed.

**Comment by kleopatra** [comment]
> 
> The specification section of the CSR should (and does) list only the spec change to `SelectionModel.isSelected`. Since there could be an impact to applications that rely on the current behavior, it also seems reasonable to list those in the Solution section (as well as the Compatibility concerns section). I think that the Summary should start by listing the spec change to `SelectionModel.isSelected`, but it seems useful to also say that the implementation of `TableXXSelectionModel` is changing as a result.

makes sense - thanks for the explanation  :)

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java (hunk 3 above)
Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableView.java:2829]: Think that I found a case where this implementation violates the invariant   `getSelectedIndices().contains(row) == isSelected(row)` -  happens in cell selection mode when we select the last column and then hide that column:

    sm.select(row, column);
    assertTrue("sanity: row " + row + "contained in selectedIndices", sm.getSelectedIndices().contains(row));
    assertTrue("sanity: row must be selected" , sm.isSelected(row));
    column.setVisible(false);
    assertTrue("after hiding column: row " + row + "contained in selectedIndices", sm.getSelectedIndices().contains(row));
    assertTrue("after hiding column: row must be selected" , sm.isSelected(row));

the last line fails for cell selection enabled with this implementation and passes when delegating to super (or not override `isSelected(int)` at all)

Aside: there is a general issue with cell selection not updated on columns modification (the selection visual is kept on the same column index without changing selectedCells accordingly - technically due to looping across the visibleLeafCells vs. all leaf columns. Might or might not be what ux requires, but the selectedCells must be in sync with the visuals always). Don't remember if we have it covered in JBS?

The complete test set:

    @Test
    public void testRowSelectionAfterSelectAndHideLastColumnMultipleCellEnabled() {
        assertRowSelectionAfterSelectAndHideLastColumn(SelectionMode.MULTIPLE, true);
    }

    @Test
    public void testRowSelectionAfterSelectAndHideLastColumnMultipleNotCellEnabled() {
        assertRowSelectionAfterSelectAndHideLastColumn(SelectionMode.MULTIPLE, false);
    }

    @Test
    public void testRowSelectionAfterSelectAndHideLastColumnSingleCellEnabled() {
        assertRowSelectionAfterSelectAndHideLastColumn(SelectionMode.SINGLE, true);
    }

    @Test
    public void testRowSelectionAfterSelectAndHideLastColumnSingleNotCellEnabled() {
        assertRowSelectionAfterSelectAndHideLastColumn(SelectionMode.SINGLE, false);
    }

    public void assertRowSelectionAfterSelectAndHideLastColumn(SelectionMode mode, boolean cellEnabled) {
        TableView<Person> table = createPersonTableView();

        TableView.TableViewSelectionModel<Person> sm = table.getSelectionModel();
        sm.setCellSelectionEnabled(cellEnabled);
        sm.setSelectionMode(mode);
        int row = 1;
        int col = table.getColumns().size() - 1;
        assertRowSelectionAfterSelectAndHideColumn(table, row, col);
    }

    private void assertRowSelectionAfterSelectAndHideColumn(TableView<Person> table, int row, int col) {
        TableViewSelectionModel<Person> sm = table.getSelectionModel();
        TableColumn<Person, ?> column = table.getColumns().get(col);
        TablePosition<Person, ?> pos = new TablePosition<>(table, row, column);

        sm.select(row, column);
        assertTrue("sanity: row " + row + "contained in selectedIndices", sm.getSelectedIndices().contains(row));
        assertTrue("sanity: row must be selected" , sm.isSelected(row));
        column.setVisible(false);
        assertTrue("after hiding column: row " + row + "contained in selectedIndices", sm.getSelectedIndices().contains(row));
        assertTrue("after hiding column: row must be selected" , sm.isSelected(row));
    }

    /**
     * Creates and returns a TableView with Persons and columns for all their properties.
     */
    private TableView<Person> createPersonTableView() {
        final ObservableList<Person> data =
                FXCollections.observableArrayList(
                        new Person("Jacob", "Smith", "jacob.smith@example.com"),
                        new Person("Isabella", "Johnson", "isabella.johnson@example.com"),
                        new Person("Ethan", "Williams", "ethan.williams@example.com"),
                        new Person("Emma", "Jones", "emma.jones@example.com"),
                        new Person("Michael", "Brown", "michael.brown@example.com"));

        TableView<Person> table = new TableView<>();
        table.setItems(data);

        TableColumn<Person, String> firstNameCol = new TableColumn("First Name");
        firstNameCol.setCellValueFactory(new PropertyValueFactory<Person, String>("firstName"));

        TableColumn<Person, String> lastNameCol = new TableColumn("Last Name");
        lastNameCol.setCellValueFactory(new PropertyValueFactory<Person, String>("lastName"));

        TableColumn<Person, String> emailCol = new TableColumn("Email");
        emailCol.setCellValueFactory(new PropertyValueFactory<Person, String>("email"));

        table.getColumns().addAll(firstNameCol, lastNameCol, emailCol);

        return table;
    }

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewMouseInputTest.java [hunk 7]
```diff
@@ -554,6 +554,7 @@ private void test_rt_38306(boolean selectTwoRows) {
         TableView.TableViewSelectionModel sm = table.getSelectionModel();
         sm.setCellSelectionEnabled(true);
         sm.setSelectionMode(SelectionMode.MULTIPLE);
+        sm.clearSelection();
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewMouseInputTest.java:558]: wondering: why did you add the clearSelection?

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewMouseInputTest.java:558]: it was an attempt to ensure a clear initial state, but you are right - it is unnecessary.

**Comment by andy-goryachev-oracle** [comment]
> the last line fails for cell selection enabled with this implementation and passes when delegating to super (or not override isSelected(int) at all)

I think you are right, @kleopatra !  Took me a while, but I see your point now.  This method should indeed delegate to super(), or not be overridden at all.

Thank you for creating the tests - they deserve to be integrated into  TableViewSelectionModelImplTest, with your permission.

**Comment by andy-goryachev-oracle** [comment]
> Aside: there is a general issue with cell selection not updated on columns modification (the selection visual is kept on the same column index without changing selectedCells accordingly - technically due to looping across the visibleLeafCells vs. all leaf columns. Might or might not be what ux requires, but the selectedCells must be in sync with the visuals always). 

Personally, I feel that any manipulations with the columns structure should clear the existing selection.  I would imagine it's such a rare operation, and the "fix" is so easy (clear selection) that it's not worth even creating an issue for - but I could be wrong.

> Don't remember if we have it covered in JBS?

I did not find a close match.  There is JDK-8095010; there is ticket you might be familiar with JDK-8093855; and one possibly related JDK-8091191.

**Comment by kevinrushforth** [comment]
I would lean towards creating a new issue, unless one of the three you listed above is a natural fit.

**Comment by kleopatra** [comment]
> 
> I think you are right, @kleopatra ! Took me a while, but I see your point now. This method should indeed delegate to super(), or not be overridden at all.
> 

I would prefer removing it (reducing clutter :) As it's a private class, there will be no publicly visible change.

> Thank you for creating the tests - they deserve to be integrated into TableViewSelectionModelImplTest, with your permission.

sure, everything here is open source and we don't want to duplicate efforts :) 

Probably need to do the same fix and test in TreeTableView selection?

**Comment by kleopatra** [comment]
> 
> Personally, I feel that any manipulations with the columns structure should clear the existing selection. I would imagine it's such a rare operation, and the "fix" is so easy (clear selection) that it's not worth even creating an issue for - but I could be wrong.
> 

haha .. are you prepared to get lynched by furious users 😜 Just imagine you selected several cells across the table, then decide to hide an unrelated column (which is not as rare as you assume :) and all your previous selection work is destroyed. 

> > Don't remember if we have it covered in JBS?
> 
> I did not find a close match. There is JDK-8095010; there is ticket you might be familiar with JDK-8093855; and one possibly related JDK-8091191.

none of those seem to be related to modification of columns, so tend to agree with Kevin that we should file a new issue

**Comment by andy-goryachev-oracle** [comment]
@kleopatra : 
good point!  I've tested with swing / JTable, and removing a column seem to deselect the cells in the removed column.  I guess this should be the expected behavior - to clear selection in the hidden column(s)?  If the app requirements call for preserving selection, this can be done on the app level (store the hidden cell selection and restore once the cells come to view)?

I'll create a separate ticket (TBD).

On another note, it seems the changes impacted/broke cell selection via mouse.  Need to investigate.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java [hunk 8]
```diff
@@ -441,7 +441,7 @@ private void updateSelection() {
         if (index == -1 || getTreeTableView() == null) return;
         if (getTreeTableView().getSelectionModel() == null) return;
 
-        boolean isSelected = getTreeTableView().getSelectionModel().isSelected(index);
+        boolean isSelected = getTreeTableView().getSelectionModel().isSelected(index, null);
         if (isSelected() == isSelected) return;
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java:445]: good example of when using the two-param method makes sense - we probably should do the same in TableRow (which does some convoluted ifs for the same result) to keep both as similar as possible. Might be done here or in a new low-priority  follow-up issue

**Comment by kleopatra** [comment]
> the tests - they deserve to be integrated into TableViewSelectionModelImplTest, with your permission.

similar tests (along with that in the bug report) adapted to treeTable should be added to the treeTable test - we should keep them as sync'ed as possible, to not make future maintainers wonder why they aren't included :)

**Comment by andy-goryachev-oracle** [comment]
> similar tests (along with that in the bug report) adapted to treeTable should be added to the treeTable test - we should keep them as sync'ed as possible, to not make future maintainers wonder why they aren't included :)

The tests were added to both TableViewSelectionModelImplTest and TreeTableViewSelectionModelImplTest.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java (hunk 8 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java:445]: unnecessary in the case of TableRow.
the change in TreeTableRow was necessary because otherwise selecting a cell by mouse selected the whole row.

**Comment by kleopatra** [comment]
> 
> The tests were added to both TableViewSelectionModelImplTest and TreeTableViewSelectionModelImplTest.

hmm .. in TreeTableViewSelectionModelImplTest, I don't see any of the  testRowSelectionAfterSelectAndHideLastColumnXX nor testSelectRowWhenInSingleCellSelectionModeIsSelected, what am I missing?

**Comment by andy-goryachev-oracle** [comment]
> hmm .. in TreeTableViewSelectionModelImplTest, I don't see any of the testRowSelectionAfterSelectAndHideLastColumnXX nor testSelectRowWhenInSingleCellSelectionModeIsSelected, what am I missing?

you are right... I am so sorry.  missed this one while cherry picking from another branch.  will fix it soon.

would it be possible to review https://github.com/openjdk/jfx/pull/858 so there is no need to cherry pick, please?

**Comment by kleopatra** [comment]
> 
> you are right... I am so sorry. missed this one while cherry picking from another branch. will fix it soon.

no problem, shit happens :)

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java (hunk 8 above)
Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java:445]: > unnecessary in the case of TableRow. 

that is .. why exactly?

> the change in TreeTableRow was necessary because otherwise selecting a cell by mouse selected the whole row.

it still does (if all cells are selected)

see me in grinning - my first guess was exactly the same as yours, but the moment I was about answering your comment (with "should behave the same and _they do_") some memory neuron fired and made me curious enough the check. The outcome is [JDK-8292353](https://bugs.openjdk.org/browse/JDK-8292353)

as that is unrelated to this fix (same before/after), we are fine here

**Review by kleopatra — APPROVED** [review]
long journey with a good end :)

Looks fine, verified tests: 
- some assertions needed to be adjusted to the fix, so they were passing before and still passing after the fix
- some where failing/passing before/after the fix

**Comment by aghaisas** [comment]
I see a very good progress on this PR. Thanks @andy-goryachev-oracle and @kleopatra !
Only thing that needs to be corrected is - the PR description has become out of date with all the review change fixes.

**Comment by andy-goryachev-oracle** [comment]
updated PR description.  thank you, @aghaisas

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java (hunk 8 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java:445]: [JDK-8292353](https://bugs.openjdk.org/browse/JDK-8292353) can be trivially fixed (PR to be filed).

**Review by kevinrushforth — APPROVED** [review]
This all looks good to me now. I left one comment on the CSR, and I'll review that once that is done.

This will give @aghaisas a chance to review as well.

**Review by aghaisas — APPROVED** [review]
Fix looks good to me!

**Comment by kleopatra** [comment]
hmm .. isn't this ready for integration? CSR being approved was the last step, or not?
